From 1f73215813c49d694856d2c3a79fa52ae803fad8 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 24 May 2019 11:00:05 +0100
Subject: [PATCH] TOOLS-2178 eng check targets should reference resources in
 deps/eng/tools Reviewed by: Trent Mick <trentm@gmail.com> Approved by: Trent
 Mick <trentm@gmail.com>

---
 CHANGES.md     |  2 ++
 Makefile       |  9 ++++++---
 deps/eng       |  2 +-
 lib/dbinit.js  | 16 +++++++++++++++-
 lib/queries.js |  4 ++--
 5 files changed, 26 insertions(+), 7 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index e9821ae..4570399 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,8 @@
 # pgstatsmon Changelog
 
 ## Not yet released
+* #26 vacuum progress function should use SETOF
+* #25 distinguish between vacuum and vacuum to prevent wraparound
 * #24 track vacuum start time
 
 ## 1.1.0
diff --git a/Makefile b/Makefile
index 8639e34..5fd08cf 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2019, Joyent, Inc. All rights reserved.
+# Copyright 2019 Joyent, Inc.
 #
 # Makefile: pgstatsmon - Postgres monitoring system
 #
@@ -54,6 +54,7 @@ TOP ?= $(error Unable to access eng.git submodule Makefiles.)
 include ./deps/eng/tools/mk/Makefile.node_modules.defs
 include ./deps/eng/tools/mk/Makefile.node_prebuilt.defs
 include ./deps/eng/tools/mk/Makefile.agent_prebuilt.defs
+include ./deps/eng/tools/mk/Makefile.smf.defs
 
 #
 # Install macros and targets
@@ -84,7 +85,8 @@ NODE_BITS_DIR	= $(PREFIX)/node
 SAPI_MANIFESTS		= pgstatsmon
 SAPI_MANIFESTS_DIRS	= $(SAPI_MANIFESTS:%=$(PREFIX)/sapi_manifests/%)
 
-SMF_MANIFESTS		= pgstatsmon metric-ports-updater
+SMF_MANIFEST_NAMES	= pgstatsmon metric-ports-updater
+SMF_MANIFESTS		= $(SMF_MANIFEST_NAMES:%=smf/manifests/%.xml)
 SMF_METHODS		= metric-ports-updater
 
 SMF_MANIFEST_DIR	= $(PREFIX)/smf/manifests
@@ -105,7 +107,7 @@ INSTALL_FILES	= $(addprefix $(PROTO), \
 		  $(BOOT_SCRIPTS:%=$(BOOT_SCRIPTS_DIR)/%) \
 		  $(SAPI_MANIFESTS_DIRS:%=%/template) \
 		  $(SAPI_MANIFESTS_DIRS:%=%/manifest.json) \
-		  $(SMF_MANIFESTS:%=$(SMF_MANIFEST_DIR)/%.xml) \
+		  $(SMF_MANIFESTS:%=$(PREFIX)/%) \
 		  $(SMF_METHODS:%=$(SMF_METHOD_DIR)/%.sh) \
 		  )
 
@@ -235,3 +237,4 @@ include ./deps/eng/tools/mk/Makefile.targ
 include ./deps/eng/tools/mk/Makefile.node_modules.targ
 include ./deps/eng/tools/mk/Makefile.node_prebuilt.targ
 include ./deps/eng/tools/mk/Makefile.agent_prebuilt.targ
+include ./deps/eng/tools/mk/Makefile.smf.targ
diff --git a/deps/eng b/deps/eng
index 126c0f0..d25b8fc 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 126c0f032b8bfddd45b14e8ee14e73e9798a013f
+Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
diff --git a/lib/dbinit.js b/lib/dbinit.js
index e86ac70..c90793a 100644
--- a/lib/dbinit.js
+++ b/lib/dbinit.js
@@ -236,6 +236,7 @@ function create_progress_vacuum_function(args, callback) {
 	query = 'DROP FUNCTION IF EXISTS get_stat_progress_vacuum(); '
 	+ 'CREATE FUNCTION public.get_stat_progress_vacuum('
 	+ 'out relname name, '
+	+ 'out vacuum_mode text, '
 	+ 'out query_start double precision, '
 	+ 'out phase bigint, '
 	+ 'out heap_blks_total bigint, '
@@ -243,9 +244,22 @@ function create_progress_vacuum_function(args, callback) {
 	+ 'out heap_blks_vacuumed bigint, '
 	+ 'out index_vacuum_count bigint, '
 	+ 'out max_dead_tuples bigint, '
-	+ 'out num_dead_tuples bigint)'
+	+ 'out num_dead_tuples bigint) '
+	+ ' RETURNS SETOF record'
 	+ ' AS $$'
 	+ ' SELECT T.relname AS relname,'
+	+ '	  CASE'
+	+ '	    WHEN'
+	+ '	      A.query ~ \'^autovacuum.*(to prevent wraparound)\''
+	+ '	    THEN \'aggressive_autovacuum\''
+	+ '	    WHEN'
+	+ '	      A.query ~ \'^autovacuum\''
+	+ '	    THEN \'autovacuum\''
+	+ '	    WHEN'
+	+ '	      A.query ~* \'^vacuum\''
+	+ '	    THEN \'manual_vacuum\''
+	+ '	    ELSE \'unknown\''
+	+ '	  END AS vacuum_mode,'
 	+ '	  EXTRACT (EPOCH FROM A.query_start) AS query_start,'
 	+ '	  S.param1+1 AS phase,'
 	+ '	  S.param2 AS heap_blks_total,'
diff --git a/lib/queries.js b/lib/queries.js
index 52e5187..b6419ae 100644
--- a/lib/queries.js
+++ b/lib/queries.js
@@ -318,7 +318,7 @@ function getQueries(config) {
 	}, {
 	     'name': 'pg_stat_progress_vacuum',
 	     'statkey': 'relname',
-	     'metadata': [ 'relname'],
+	     'metadata': [ 'relname', 'vacuum_mode' ],
 	     'sql': [
 	         'SELECT * FROM get_stat_progress_vacuum()'
 	     ].join('\n'),
@@ -327,7 +327,7 @@ function getQueries(config) {
 	           'vacuum', 'expires': true, 'expiryPeriod': expiryPeriod },
 	         { 'attr': 'query_start', 'help': 'unix epoch timestamp of ' +
 	           'the vacuum began', 'expires': true,
-		    'expiryPeriod': expiryPeriod },
+	           'expiryPeriod': expiryPeriod },
 	         { 'attr': 'heap_blks_total', 'help': 'total number of heap ' +
 	           'blocks in the table as of the beginning of the scan',
 	           'expires': true, 'expiryPeriod': expiryPeriod },
-- 
2.21.0

